code/modules/mob/living/silicon/pai/software.dm

```diff
@@ -376,6 +397,45 @@
 	return dat
 
 
+/mob/living/silicon/pai/proc/uninstallSoftware()
+	if(!uninstall_lock) //currently not uninstalling software
+		var/dat = ""
+		dat += {"<h2>CentComm pAI Module Local Software</h2><br>
+			<pre>Remaining Available Memory: [ram]</pre><br>
+			<p style=\"text-align:center\"><b>Currently installed modules.</b><br>"}
+		for(var/s in software)
+			var/cost = available_software[s]
+			var/displayName = uppertext(s)
+			if(!cost)
+				continue
+			dat += "<a href='byond://?src=\ref[src];software=uninstall;sub=1;uninstall=[s]'>[displayName]</a> ([cost]) <br>"
+			dat += "<font color='blue'>Estimated Uninstall Time: ([cost*2]) seconds: </font><br>"
+		dat += "<font color='red'> WARNING: The uninstallation process must complete before allocated RAM can be released. During uninstallation no further sofware can be uninstalled. The duration is typically 2 seconds per unit RAM. </font>"
+		dat += "</p>"
+		return dat
+	//during an uninstall.
+	else
+		var/dat = {"<h3>Uninstalling</h3>: "}
+		dat += "... [uninstallprogress]% complete.<br>"
+		dat += "<a href='byond://?src=\ref[src];software=uninstall;cancel=1;sub=0'>Cancel</a> <br>"
+		return dat
+
+/mob/living/silicon/pai/proc/uninstall_loop(var/cost)
+	uninstall_lock = TRUE
+	var/duration = cost*2 //2 seconds per unit RAM
+	while(uninstallprogress < 100)
+		uninstallprogress += round((1/(duration))*100)  //store and calculate the value in percentages. easier to display.
+		uninstallprogress = min(100,uninstallprogress) //Never go above 100
+		if(!uninstall_lock) //if canceled by the PAI, this becomes FALSE and the loop returns false here
+			uninstallprogress =0
+			return 0
+		if(uninstallprogress >= 100)
+			uninstallprogress = 0
+			to_chat(src, "<span class='notice'>Software uninstallation complete!</span>")
+			uninstall_lock = FALSE
+			return 1
+		sleep(10)
```

Comment:
1 SECONDS

Reply:
I'm not following

Reply:
just a DEFINE you can use to make your code more readable, instead of sleep(10) use sleep(1 SECONDS)
this does the same thing but makes it more obvious how long you intended it to be

in your application it's obvious (to coders who understand deciseconds), but this really comes in handy for sleep(1 HOURS) instead of sleep(36000)

Reply:
understood. I'll do the same to the wirejack  (I copy pasted that line from wirejack)